src/binary/paths.ts

```diff
@@ -5,68 +5,103 @@ import {
   BINARY_UPDATE_URL,
   BINARY_UPDATE_VERSION_FILE_URL,
 } from "../globals/consts";
+import { ONPREM } from "../onPrem";
 
-let binaryRootPath: string | undefined;
+let BINARY_ROOT_PATH: string | undefined;
 const ARCHITECTURE = getArch();
 const SUFFIX = getSuffix();
 const BUNDLE_SUFFIX = getBundleSuffix();
 
 export async function setBinaryRootPath(
   extensionContext: vscode.ExtensionContext
 ): Promise<void> {
-  binaryRootPath =
+  if (ONPREM) {
+    const base = `binaries/${getArch()}-${getPlatform()}`;
+    BINARY_ROOT_PATH = path.join(extensionContext.extensionPath, base);
+    await makeExecutable(BINARY_ROOT_PATH);
+    return;
+  }
+  BINARY_ROOT_PATH =
     extensionContext.extensionMode === vscode.ExtensionMode.Test
       ? path.join(__dirname, "..", "..", "binaries")
       : path.join(extensionContext.globalStorageUri.fsPath, "binaries");
 
   try {
-    await fs.mkdir(binaryRootPath, { recursive: true });
+    await fs.mkdir(BINARY_ROOT_PATH, { recursive: true });
   } catch (err) {
     // Exception is thrown if the path already exists, so ignore error.
   }
 }
 
+export async function makeExecutable(binaryRootPath: string) {
+  const bundleDirectory = binaryRootPath;
+  if (process.platform === "win32") {
+    return;
+  }
+  const files = await fs.readdir(bundleDirectory);
+  console.log("making files executable", files);
+  await Promise.all(
+    files.map((file) => fs.chmod(path.join(bundleDirectory, file), 0o755))
+  );
+}
+
+export function bundledTabnineBinaryPath(): string {
+  return path.join(<string>BINARY_ROOT_PATH, binaryName());
```

Comment:
BINARY_ROOT_PATH can be undefined

Reply:
this is the current convention. if it could happen now, it could happen before. the design isn't great to have dangling references

Reply:
i meant the casting you are doing here, it will be more clear to adjust the return type and this casting